### PR TITLE
Refactors credentials to be class-based

### LINKED ISSUE #207: User directory for configuration
Right now, the configuration file for saving credentials is located in the same place as the cartoframes library https://github.com/CartoDB/cartoframes/blob/master/cartoframes/credentials.py#L8-L9.

This will cause permission issues with global installs (`pip install` without virtual env). My suggestion would be to use the os-dependent configuration directory. A good way to do this is to use the [appdirs module](https://pypi.python.org/pypi/appdirs/1.4.3) function, `user_config_dir`. It also has the nice thing that this is per-user instead of per-system. The bad part is yet another dependency.

cc @andy-esch @stuartlynn

### LINKED ISSUE #208: Class for credentials
This is just a suggestion. I feel that the credentials handling code is a bit cumbersome to use, with a lot of little functions for saving each setting. Although this works, it will become a mess if we add more settings.

My suggestion:
- Credentials class
- Pass the configuration file in the constructor (if not set, have some default configuration path)
- One attribute per setting (`api_key`, `base_url`)
- One method to `load` and a second one to `save` configuration. Probably calling `load` in the constructor, and `save` in the attribute setters.
- Allow passing a `Credentials` instance to the CF context. This way, all settings are encapsulated with a single parameter.

This is my very rough idea. I think it's worth it for future extensibility as more settings may very well be needed. What do you think? @andy-esch @stuartlynn

### PR DESCRIPTION
**Pull request by andy-esch**
Credentials are now class-based instead of a module of functions. This cleans up the code quite a bit (reduction of around 120 lines from the source code), and much easier logic.

TODO:

- [x] context.CartoContext should accept a `Credentials` parameter

Closes #207 and closes #208 

cc @javitonino

### CODE REVIEW

File: cartoframes/credentials.py [hunk 1]
```diff
@@ -1,208 +1,127 @@
...
+            else:
+                self._base_url = 'https://{}.carto.com/'.format(self._username)
+        elif cred_file:
+            self._retrieve(cred_file)
+        else:
+            try:
+                creds = json.loads(open(_DEFAULT_PATH).read())
+                self._key = creds['key']
+                self._base_url = creds['base_url']
+                self._username = creds.get('username')
```
Comment (javitonino) [inline cartoframes/credentials.py]: This should probably just be `self._retrieve()`, maybe with some error handling.

File: cartoframes/credentials.py [hunk 2]
```diff
@@ -1,208 +1,127 @@
...
+                cloud-based account. Generally of the form
+                ``https://your_user_name.carto.com/`` for cloud-based accounts.
+                If on-prem or otherwise, contact your admin.
+        """
+        self.__init__(key=(key or self._key),
+                      username=(username or self._username),
+                      base_url=base_url)
+
+    def key(self, key=None):
+        """Return API key if no arguemnt is passed. Updates key to `key`
```
Comment (javitonino) [inline cartoframes/credentials.py]: Typo: `arguemnt`

File: cartoframes/credentials.py (hunk 1 above)
Comment (andy-esch) [reply cartoframes/credentials.py]: Totally, good catch!

File: cartoframes/credentials.py (hunk 2 above)
Comment (andy-esch) [reply cartoframes/credentials.py]: 👍

**Pull request merged**